Update SVCB

DesWurstes opened the pull request:
The only breaking change worth mentioning is renaming SVCB_ECHCONFIG to SVCB_ECH. We can avoid this and use its legacy name internally if you wish.

I had chosen ( https://github.com/miekg/dns/pull/1067#issuecomment-699635130 ) to forbid IPv4 addresses in ipv6hint to be consistent with the rest of the library. I don't know why ( https://github.com/MikeBishop/dns-alt-svc/issues/361 ) but the SVCB spec requires otherwise.

@miekg @tmthrgd Would you like to review?

miekg commented:
Breaking changes in this record I fine, it's still draft IIRC

> I had chosen ( https://github.com/miekg/dns/pull/1067#issuecomment-699635130 ) to forbid IPv4 addresses in ipv6hint to be consistent with the rest of the library. I don't know why ( https://github.com/MikeBishop/dns-alt-svc/issues/361 ) but the SVCB spec requires otherwise.

That sgtm, internal consistency is a good thing to have

miekg commented on svcb.go:
```diff
@@ -13,14 +13,14 @@ import (
-// Keys defined in draft-ietf-dnsop-svcb-https-01 Section 12.3.2.
+// Keys defined in draft-ietf-dnsop-svcb-https-08 Section 14.3.2.
 const (
 	SVCB_MANDATORY       SVCBKey = 0
```
```suggestion
	SVCB_MANDATORY       SVCBKey = iota
```

miekg commented on svcb.go:
```diff
@@ -13,14 +13,14 @@ import (
 const (
 	SVCB_MANDATORY       SVCBKey = 0
 	SVCB_ALPN            SVCBKey = 1
 	SVCB_NO_DEFAULT_ALPN SVCBKey = 2
 	SVCB_PORT            SVCBKey = 3
 	SVCB_IPV4HINT        SVCBKey = 4
-	SVCB_ECHCONFIG       SVCBKey = 5
+	SVCB_ECH             SVCBKey = 5
 	SVCB_IPV6HINT        SVCBKey = 6
```
can remove all the "= N" and let `iota` do its thing

miekg replied:
think you can even remove the type in that case

miekg commented on svcb.go:
```diff
@@ -13,14 +13,14 @@ import (
-	SVCB_ECHCONFIG       SVCBKey = 5
+	SVCB_ECH             SVCBKey = 5
 	SVCB_IPV6HINT        SVCBKey = 6
 	svcb_RESERVED        SVCBKey = 65535
```
```suggestion

	svcb_RESERVED        SVCBKey = 65535
```

miekg commented on svcb.go:
```diff
@@ -200,12 +200,14 @@ func makeSVCBKeyValue(key SVCBKey) SVCBKeyValue {
 	Target   string         `dns:"domain-name"`
-	Value    []SVCBKeyValue `dns:"pairs"` // Value must be empty if Priority is zero.
+	Value    []SVCBKeyValue `dns:"pairs"`
+	// If Priority is zero, Value should be empty.
```
move this as a comment after Priority

miekg commented on svcb.go:
```diff
@@ -200,12 +200,14 @@ func makeSVCBKeyValue(key SVCBKey) SVCBKeyValue {
-	Value    []SVCBKeyValue `dns:"pairs"` // Value must be empty if Priority is zero.
+	Value    []SVCBKeyValue `dns:"pairs"`
+	// If Priority is zero, Value should be empty.
+	// If it is not, when receiving, Value must be considered empty.
```
not sure what this adds, expand/delete?

miekg commented on svcb.go:
```diff
@@ -235,15 +237,26 @@ type SVCBKeyValue interface {
+//
+// It is incumbent upon the user of this library to reject the RRSet if
+// or avoid constructing such an RRSet that:
+// - "mandatory" is included as one of the keys of mandatory
```
if you want this to make a list in godoc you need empty lines before and after. Maybe just make this more prose?

DesWurstes replied:
I can't turn this list into a paragraph; it would be unreadable.

miekg commented on svcb.go:
```diff
@@ -235,15 +237,26 @@ type SVCBKeyValue interface {
 //
 //	s := &dns.SVCB{Hdr: dns.RR_Header{Name: ".", Rrtype: dns.TypeSVCB, Class: dns.ClassINET}}
 //	e := new(dns.SVCBMandatory)
 //	e.Code = []uint16{65403}
```
this is just a random number in the "correct" range?

miekg commented on svcb.go:
```diff
@@ -302,6 +315,7 @@ func (s *SVCBMandatory) copy() SVCBKeyValue {
 
 // SVCBAlpn pair is used to list supported connection protocols.
+// The user of this library must ensure that at least one protocol is listed when alpn is present.
 // Protocol ids can be found at:
```
IDs

miekg commented on svcb.go:
```diff
@@ -370,6 +384,7 @@ func (s *SVCBAlpn) copy() SVCBKeyValue {
 }
 
 // SVCBNoDefaultAlpn pair signifies no support for default connection protocols.
+// Should be used in conjunction with alpn.
```
'It should..."

and then "only be used" ?

miekg commented on svcb.go:
```diff
@@ -522,42 +537,42 @@ func (s *SVCBIPv4Hint) copy() SVCBKeyValue {
 //	h.Hdr = dns.RR_Header{Name: ".", Rrtype: dns.TypeHTTPS, Class: dns.ClassINET}
-//	e := new(dns.SVCBECHConfig)
+//	e := new(dns.SVCBECH)
 //	e.ECH = []byte{0xfe, 0x08, ...}
```
doc doesn't match struct members anymore

miekg commented on svcb.go:
```diff
@@ -522,42 +537,42 @@ func (s *SVCBIPv4Hint) copy() SVCBKeyValue {
-type SVCBECHConfig struct {
-	ECH []byte
+type SVCBECH struct {
+	ECHConfigList []byte // includes the redundant length prefix
```
this is too long.  Just `Config`?  What does the draft use?

DesWurstes replied:
The draft uses `ECHConfigList`

DesWurstes commented:
@miekg  Many thanks for your review. I have addresses your comments except "if you want this to make a list in godoc you need empty lines before and after. Maybe just make this more prose?" because I don't know how to turn this into a paragraph without making it unreadable.

> > I had chosen ( https://github.com/miekg/dns/pull/1067#issuecomment-699635130 ) to forbid IPv4 addresses in ipv6hint to be consistent with the rest of the library. I don't know why ( https://github.com/MikeBishop/dns-alt-svc/issues/361 ) but the SVCB spec requires otherwise.
>
> That sgtm, internal consistency is a good thing to have

On the SVCB side, they decided to update their example, so this library now achieves full compliance and consistency. Hence I updated the relevant code.

tmthrgd commented on svcb.go:
```diff
@@ -167,10 +168,10 @@ func (rr *SVCB) parse(c *zlexer, o string) *ParseError {
 	}
+
+	// Don't check rr.Priority == 0 && len(xs) > 0
+	// as explained in parse_test.go
```
Other way around. Test can reference comment in the code, not the opposite.

tmthrgd reviewed (changes requested):
Most of the change is fine, but I really don't think we can or should make backwards incompatible changes. Nothing here was marked as WIP or anything. Considering the name changes don't have any actual impact on the code, I think they should be left as-is regardless of whether that differs from the draft spec or not. I'm fine with the other changes though.

DesWurstes commented on svcb.go:
```diff
@@ -13,16 +13,17 @@ import (
 // SVCBKey is the type of the keys used in the SVCB RR.
 type SVCBKey uint16
 
-// Keys defined in draft-ietf-dnsop-svcb-https-01 Section 12.3.2.
```
@tmthrgd Does this count?

tmthrgd replied:
No it doesn't.

DesWurstes commented:
Thanks for the review. I addressed the issues.

tmthrgd reviewed (approved):
LGTM.

DesWurstes commented:
Hello again @miekg . I have been impatiently waiting for Cloudflare to implement ECH over SVCB so it would be great if we could get this merged unless further reviews are planned. Many thanks.

DesWurstes commented:
👍thank you
